Wagtail 4.2 new incompatibility with dj-inmemorystorage

<!--
Found a bug? Please fill out the sections below. 👍
-->

### Issue Summary

I tried today to upgrade to Wagtail 4.2 and noticed that my image related tests
stopped working. I use [dj-inmemorystorage](https://github.com/waveaccounting/dj-inmemorystorage)
to speed up my file base tests a little bit and it seems like it's incompatible with willow.

### Steps to Reproduce

1. Install `pip install dj-inmemorystorage`
2. Add `DEFAULT_FILE_STORAGE = "inmemorystorage.InMemoryStorage"` to your test settings
3. Running a test creating an Image model will now raise an `willow.image.UnrecognisedImageFormatError` error see below for an example

```python
@pytest.mark.django_db
def test_create_image_from_scratch():
    from django.core.files.uploadedfile import SimpleUploadedFile
    from wagtail.images.models import Image
    # This is a 1x1 black png
    png = (
        b"\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00"
        b"\x00\x01\x00\x00\x00\x01\x08\x06\x00\x00\x00"
        b"\x1f\x15\xc4\x89\x00\x00\x00\rIDATx\x9cc````"
        b"\x00\x00\x00\x05\x00\x01\xa5\xf6E@\x00\x00"
        b"\x00\x00IEND\xaeB`\x82"
    )

    simple_png = SimpleUploadedFile(name="test.png", content=png, content_type="image/png")
    image = Image(file=simple_png)
    image.save()
```

<details>
<summary>Here's the traceback</summary>

```python
╰─>$ pytest -k test_create_image_from_scratch  1283ms  Di  7 Feb 19:21:39 2023
============================= test session starts =============================
platform darwin -- Python 3.11.1, pytest-7.2.1, pluggy-1.0.0
Using --randomly-seed=1234
django: settings: tests.settings (from ini)
rootdir: /Users/jochen/projects/django-cast, configfile: pyproject.toml, testpaths: tests
plugins: Faker-16.6.1, randomly-3.12.0, cov-4.0.0, django-4.5.2, anyio-3.6.2
collected 207 items / 206 deselected / 1 selected

tests/wagtail_image_views_test.py F                                     [100%]

================================== FAILURES ===================================
_______________________ test_create_image_from_scratch ________________________

    @pytest.mark.django_db
    def test_create_image_from_scratch():
        from django.core.files.uploadedfile import SimpleUploadedFile
        from wagtail.images.models import Image
        # This is a 1x1 black png
        png = (
            b"\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00"
            b"\x00\x01\x00\x00\x00\x01\x08\x06\x00\x00\x00"
            b"\x1f\x15\xc4\x89\x00\x00\x00\rIDATx\x9cc````"
            b"\x00\x00\x00\x05\x00\x01\xa5\xf6E@\x00\x00"
            b"\x00\x00IEND\xaeB`\x82"
        )

        simple_png = SimpleUploadedFile(name="test.png", content=png, content_type="image/png")
        image = Image(file=simple_png)
>       image.save()

tests/wagtail_image_views_test.py:19:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/base.py:812: in save
    self.save_base(
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/base.py:863: in save_base
    updated = self._save_table(
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/base.py:1006: in _save_table
    results = self._do_insert(
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/base.py:1047: in _do_insert
    return manager._insert(
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/manager.py:85: in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/query.py:1791: in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/sql/compiler.py:1659: in execute_sql
    for sql, params in self.as_sql():
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/sql/compiler.py:1583: in as_sql
    value_rows = [
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/sql/compiler.py:1584: in <listcomp>
    [
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/sql/compiler.py:1585: in <listcomp>
    self.prepare_value(field, self.pre_save_val(field, obj))
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/sql/compiler.py:1533: in pre_save_val
    return field.pre_save(obj, add=True)
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/fields/files.py:316: in pre_save
    file.save(file.name, file.file, save=False)
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/fields/files.py:93: in save
    setattr(self.instance, self.field.attname, self.name)
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/fields/files.py:378: in __set__
    self.field.update_dimension_fields(instance, force=True)
../../.virtualenvs/cast/lib/python3.11/site-packages/django/db/models/fields/files.py:490: in update_dimension_fields
    width = file.width
../../.virtualenvs/cast/lib/python3.11/site-packages/django/core/files/images.py:20: in width
    return self._get_image_dimensions()[0]
../../.virtualenvs/cast/lib/python3.11/site-packages/wagtail/images/models.py:201: in _get_image_dimensions
    self._dimensions_cache = self.get_image_dimensions()
../../.virtualenvs/cast/lib/python3.11/site-packages/wagtail/images/models.py:212: in get_image_dimensions
    image = willow.Image.open(self)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'willow.image.Image'>
f = <WagtailImageFieldFile: original_images/test.png>

    @classmethod
    def open(cls, f):
        # Detect image format
        image_format = imghdr.what(f)

        # Find initial class
        initial_class = INITIAL_IMAGE_CLASSES.get(image_format)
        if not initial_class:
            if image_format:
                raise UnrecognisedImageFormatError("Cannot load %s images" % image_format)
            else:
>               raise UnrecognisedImageFormatError("Unknown image format")
E               willow.image.UnrecognisedImageFormatError: Unknown image format

../../.virtualenvs/cast/lib/python3.11/site-packages/willow/image.py:97: UnrecognisedImageFormatError
=========================== short test summary info ===========================
FAILED tests/wagtail_image_views_test.py::test_create_image_from_scratch - willow.image.UnrecognisedImageFormatError: Unknown image format
====================== 1 failed, 206 deselected in 0.40s ======================
```

</details>

Any other relevant information. For example, why do you consider this a bug and what did you expect to happen instead?

I don't know if this is a bug in Wagtail. The `dj-inmemorystorage` package looks pretty stale maybe it's using
an outdated version of the Django storage API? On the other hand: With Wagtail 4.1.2 everything worked fine.

-   I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: (yes / no)
yes

### Technical details

-   Python version: 3.11.1
-   Django version: 4.1.6
-   Wagtail version: 4.2.0
-   Browser version: no browser :)

--- a/wagtail/images/models.py
+++ b/wagtail/images/models.py
@@ -214,6 +214,8 @@ class WagtailImageFieldFile(models.fields.files.ImageFieldFile):
         finally:
             if close:
                 self.close()
+            else:
+                self.seek(0)
 
 
 class WagtailImageField(models.ImageField):

--- a/wagtail/test/dummy_external_storage.py
+++ b/wagtail/test/dummy_external_storage.py
@@ -5,6 +5,7 @@
 #  - Calling .path on the storage or image file raises NotImplementedError
 #  - File.open() after the file has been closed raises an error
 #  - File.size exceptions raise DummyExternalStorageError
+#  - Storage._save() fails loudly if the content file's pointer is not at the start
 
 from django.core.files.base import File
 from django.core.files.storage import FileSystemStorage, Storage
@@ -26,10 +27,15 @@ class DummyExternalStorage(Storage):
         # File object
         return DummyExternalStorageFile(open(self.wrapped.path(name), mode))
 
-    # Wrap all other functions
-
     def _save(self, name, content):
+        file_pos = content.tell()
+        if file_pos != 0:
+            raise ValueError(
+                "Content file pointer should be at 0 - got %d instead" % file_pos
+            )
         return self.wrapped._save(name, content)
+
+    # Wrap all other functions
 
     def delete(self, name):
         self.wrapped.delete(name)

--- a/wagtail/tests/test_tests.py
+++ b/wagtail/tests/test_tests.py
@@ -1,9 +1,11 @@
 import json
 
+from django.core.files.uploadedfile import SimpleUploadedFile
 from django.test import TestCase
 
 from wagtail.admin.tests.test_contentstate import content_state_equal
 from wagtail.models import PAGE_MODEL_CLASSES, Page, Site
+from wagtail.test.dummy_external_storage import DummyExternalStorage
 from wagtail.test.testapp.models import (
     BusinessChild,
     BusinessIndex,
@@ -429,3 +431,28 @@ class TestFormDataHelpers(TestCase):
     def test_rich_text_with_alternative_editor(self):
         result = rich_text("<h2>title</h2><p>para</p>", editor="custom")
         self.assertEqual(result, "<h2>title</h2><p>para</p>")
+
+
+class TestDummyExternalStorage(WagtailTestUtils, TestCase):
+    def test_save_with_incorrect_file_object_position(self):
+        """
+        Test that DummyExternalStorage correctly warns about attempts
+        to write files that are not rewound to the start
+        """
+        # This is a 1x1 black png
+        png = (
+            b"\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00"
+            b"\x00\x01\x00\x00\x00\x01\x08\x06\x00\x00\x00"
+            b"\x1f\x15\xc4\x89\x00\x00\x00\rIDATx\x9cc````"
+            b"\x00\x00\x00\x05\x00\x01\xa5\xf6E@\x00\x00"
+            b"\x00\x00IEND\xaeB`\x82"
+        )
+        simple_png = SimpleUploadedFile(
+            name="test.png", content=png, content_type="image/png"
+        )
+        simple_png.read()
+        with self.assertRaisesMessage(
+            ValueError,
+            "Content file pointer should be at 0 - got 70 instead",
+        ):
+            DummyExternalStorage().save("test.png", simple_png)

